### ISSUE TITLE
Check GitHub Actions behaviour when pulling a Docker image in a shell step

### ISSUE THREAD

**Issue by weibeld** [weibeld/cloudnativeindex#7]
A different approach for running the validation scripts would be to have them start a container and execute their operations in this container themselves, instead of running the entire GitHub Actions job in a container.

However, the question is whether the container image would in this case be cached by GitHub Actions between executions of the job (as it is the case when running an entire job in a container). If yes, then it would be a viable approach to make the validations exactly the same on GitHub Actions and locally executed by the pre-commit hook because they would then run in the same container. However, if the container images are not cached, then it's probably not worth it, as the image would need to be pulled on every execution of the GitHub Actions job, which is very inefficient.

**Comment by weibeld** [comment]
Experimentation results:
- GitHub Actions does **not** cache pulled Docker images between runs, but pulls them every time.
- However, [GitHub-hosted runners](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners) have some commonly used Docker images pre-cached (for example, see [here](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-Readme.md) for the Ubuntu runner), so these images are never pulled.
  - This is not only useful for precisely these pre-cached images, but also for  images that are base on these pre-cached images. For example, if an image is based on `ubuntu:20.04` (which is pre-cached), then only the layers added on top of `ubuntu:20.04` have to be pulled and all other layers can be used from the pre-cached `ubuntu:20.04` image.
  - This also extends to variations of the pre-cached images. For example, if an image is based on `ubuntu:22.04` (which is not pre-cached), then still a large part of the data can be used from the cached `ubuntu:20.04` image, because `ubuntu:22.04` still shares some layers with `ubuntu:20.04`.
- Pulling images from Docker Hub is quite fast (about 30 sec for a 400 MB image).
- There are various approaches to manually cache Docker images after a first pull (see [this overview](https://dev.to/dtinth/caching-docker-builds-in-github-actions-which-approach-is-the-fastest-a-research-18ei)). The most obvious one is manually saving a pulled image as a tar archive with `docker save` and on subsequent runs loading it with `docker load` instead of pulling it. However, `docker save` and `docker load` seem to be very slow and `docker load` may take as much or even more time than simply pulling the image.

In summary, there seems to be no elegant way of caching pulled Docker images between GitHub Actions job runs. The approaches that exist probably provide little gain with respect to their complexity/ugliness.

That being said, the best approach is probably to forego caching and rely on the fast Docker image pull speed of GitHub Actions, and especially on the [pre-cached Docker images](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-Readme.md#cached-docker-images) on the GitHub-hosted runners. That means, wherever possible, use Docker images that are based on one of these pre-cached images.
